### PR TITLE
Add implementation of `spectrogram_batch`

### PR DESCRIPTION
**Pull request by ravenouse**
# What does this PR do?

This pull request introduces the implementation of `spectrogram_batch`, specifically optimized for batch processing through broadcasting techniques. The primary goal is to reduce the data processing time during feature extraction, which is a critical step in working with audio models like `whisper`.

### Motivation
In my work and research with the whisper model, I observed that the feature extraction step can be exceedingly time-consuming, taking up to 10 hours for certain audio datasets. <br>
In my opinion, the bottleneck is primarily due to the lack of batch processing support in the current `spectrogram`  and `FeatureExtractor` implementations, resulting in iterative calls within a for-loop, as illustrated below:
```Python
# Reference: https://github.com/huggingface/transformers/blob/main/src/transformers/models/whisper/feature_extraction_whisper.py#L250
input_features = [self._np_extract_fbank_features(waveform) for waveform in input_features[0]]
```

### Future Work
The current branch only adds a basic implementation of the `spectrogram_batch`. 
To bring this implementation to production level, I believe there are several steps needed to be done:
1. **Extensive Testing**: Implementing a comprehensive suite of tests to evaluate the function’s performance and correctness across different parameter settings. I only tested the implementation in the whisper models' setting. 
2. **Integration**: Modifying existing feature extractor codes to incorporate the new batch processing function.

I am fully committed to continuing the development and testing of this feature. However, given the extensive changes and the potential impact on the library, I am reaching out for collaboration and support from the community/maintainers. Any guidance, suggestions, or contributions to this effort would be immensely appreciated.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sanchit-gandhi
@ArthurZucker

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
Hey! This looks promising and I think would be welcome! We had to go back to using `torchaudio` for some processing specifically because of the overhead. 
I think something to look forward too is also chunking single long file to batches for even faster pre-processings! 

FYI @ylacombe

**Comment by sanchit-gandhi** [comment]
Super cool PR @ravenouse, a most welcome upgrade! As mentioned by @ArthurZucker, one other optimisation we can employ is using the `torch.stft` backend for computing the spectrograms: https://github.com/huggingface/transformers/pull/26119#issue-1892916324. This yields approx 4x speed-up for processing with bs=1. I believe @ylacombe is considering adding support for this, which could be a nice parallel to your PR!

**Comment by ylacombe** [comment]
Thanks for working on this @ravenouse, this looks super promising and will clearly be valuable for any audio-related model!

I'm indeed considering adding a torch alternative of this numpy implementation! What would be really good for these current/future improvements is that we conduct extensive speed benchmark to allow users to make an informed choice when choosing implementation!

**Comment by ArthurZucker** [comment]
cc @ylacombe and @sanchit-gandhi let's try to maybe merge this now and have the rest as a todo?

**Review by ylacombe — COMMENTED** [review]
Hey @ravenouse,
First of all, thank you for taking such a useful initiative, and for your time and effort, it's already a great PR! Also sorry for taking such a long time to review!

Before going any further, I think we need to discuss several points!

As you rightfully pointed out, the current implementation process sample one-by-one resulting in a sub-optimal speed.  
However and at the moment, your current implementation expects and works with a batch of **similar size** inputs.  But how to deal with inputs of different sizes ?

From your code, I'm under the impression that we first have to pad the inputs, then post-process outside of the function to get the right results. Is that correct?
In that case, does it give the expected results? In other words, do we get the same results by processing the samples one-by-one versus processing the samples by batch? With your current implementation and **tests**, we cannot know!


IMO, we should do this pre and post-processing directly in the `spectrogram_batch`. So I would expect:
- `spectrogram_batch` to take as input a list of numpy array (i.e a list of 1D audio waveform) and to output a list of spectrogram i.e a list of different-size spectrograms
-  `spectrogram_batch` to pad inputs, then do batch processing (while keeping track of the output sizes), and finally to output list of different-size spectrograms by truncating spectrograms one by one according to the corresponding output size.
- **tests that make sure** that inputs (random or not) with different size passed to `spectrogram_batch` to get the same results as compared to pass process them one-by-one

Once that's done, we'll be sure that the function behaves as expected!

Again, thank you for your effort here, let me know if I can be of any help here!

**Comment by ylacombe** [comment]
Hey @ravenouse, thanks for all the progress so far and happy new year! Could you actually update the code here so that it's easier to review, test and keep track of the comments?
Many thanks!

**Comment by sanchit-gandhi** [comment]
Gently pining @ylacombe

File: src/transformers/audio_utils.py [hunk 1]
```diff
@@ -476,6 +476,142 @@ def spectrogram(
...
+    mel_filters: Optional[np.ndarray] = None,
+    mel_floor: float = 1e-10,
+    log_mel: Optional[str] = None,
+    reference: float = 1.0,
+    min_value: float = 1e-10,
+    db_range: Optional[float] = None,
+    remove_dc_offset: Optional[bool] = None,
+    dtype: np.dtype = np.float32,
+) -> List[np.ndarray]:
+    window_length = len(window)
```
Comment (ylacombe) [inline src/transformers/audio_utils.py:501]: Let's copy and adapt [the docstrings](https://github.com/huggingface/transformers/blob/bb4f816ad4993a5ed15f8cfd7dae67573c88e1d7/src/transformers/audio_utils.py#L304) from`spectrogram` !

File: src/transformers/audio_utils.py [hunk 2]
```diff
@@ -476,6 +476,142 @@ def spectrogram(
...
+    if center:
+        padding = [(int(frame_length // 2), int(frame_length // 2))]
+        padded_waveform_list = [
+            np.pad(
+                waveform,
+                padding,
+                mode=pad_mode,
+            )
+            for waveform in waveform_list
+        ]
```
Comment (ylacombe) [inline src/transformers/audio_utils.py:535]: Out of curiosity, shouldn't we have `padding` adapted for each waveform of the list ?

File: src/transformers/audio_utils.py [hunk 3]
```diff
@@ -476,6 +476,142 @@ def spectrogram(
...
+
+    # Batch pad the waveform
+    max_length = max(original_waveform_lengths)
+    padded_waveform_batch = np.array(
+        [
+            np.pad(waveform, (0, max_length - len(waveform)), mode="constant", constant_values=0)
+            for waveform in padded_waveform_list
+        ],
+        dtype=dtype,
+    )
```
Comment (ylacombe) [inline src/transformers/audio_utils.py:548]: Too bad we don't have a numpy equivalent to [`torch.nn.utils.rnn.pad_sequence`](https://pytorch.org/docs/stable/generated/torch.nn.utils.rnn.pad_sequence.html)

File: src/transformers/audio_utils.py [hunk 4]
```diff
@@ -476,6 +476,115 @@ def spectrogram(
...
+    if power is not None and log_mel is not None:
+        if log_mel == "log":
+            spectrogram = np.log(spectrogram)
+        elif log_mel == "log10":
+            spectrogram = np.log10(spectrogram)
+        elif log_mel == "dB":
+            if power == 1.0:
+                spectrogram = amplitude_to_db(spectrogram, reference, min_value, db_range)
+            elif power == 2.0:
+                spectrogram = power_to_db(spectrogram, reference, min_value, db_range)
```
[the remark this answers is not in the document]
Comment (ylacombe) [reply src/transformers/audio_utils.py:602]: Did you have time to look into this ? Seems unresolved

File: src/transformers/audio_utils.py [hunk 5]
```diff
@@ -476,6 +476,142 @@ def spectrogram(
...
+            else:
+                raise ValueError(f"Cannot use log_mel option '{log_mel}' with power {power}")
+        else:
+            raise ValueError(f"Unknown log_mel option: {log_mel}")
+
+    spectrogram = np.asarray(spectrogram, dtype)
+
+    spectrogram_list = [spectrogram[i, : true_num_frames[i], :].T for i in range(len(true_num_frames))]
```
Comment (ylacombe) [inline src/transformers/audio_utils.py:610]: Great!

File: tests/utils/test_audio_utils.py
Comment (ylacombe) [inline tests/utils/test_audio_utils.py:1]: I've got the exact same comment than last time:

> * **tests that make sure** that inputs (random or not) with different size passed to `spectrogram_batch` to get the same results as compared to pass process them one-by-one

We want waveforms of different lengths to produce the same results, whether transmitted in batches or one at a time.

**Review by ylacombe — COMMENTED** [review]
Hey @ravenouse, again sorry for the wait!

This is definitely a great improvement and looks really great! 
Your modeling code looks great, I'd focus in the next steps on:
- adapting the tests to your new implementation. Remember that we want to test that different waveforms with different lengths produce the same results in batch or sequentially.
-  add docstrings

Testing extensively will also allow to pinpoint some issues that you and I may have overlooked while writing/reviewing your code (e.g I'm not totally sure it works with amplitude and power spectrogram!

Again, thanks for the great work!

**Comment by ylacombe** [comment]
Hey @ravenouse, of course!

**Comment by ylacombe** [comment]
Hey @ravenouse, thanks for updating the code and for the Colab, I'll take a look ASAP!

**Comment by amyeroberts** [comment]
Gentle ping @ylacombe

**Comment by amyeroberts** [comment]
Another ping @ylacombe

File: src/transformers/audio_utils.py [hunk 6]
```diff
@@ -631,6 +839,55 @@ def power_to_db(
...
+
+    spectrogram = np.clip(spectrogram, a_min=min_value, a_max=None)
+    spectrogram = 10.0 * (np.log10(spectrogram) - np.log10(reference))
+
+    if db_range is not None:
+        if db_range <= 0.0:
+            raise ValueError("db_range must be greater than zero")
+        # Apply db_range clipping per batch item
+        max_values = spectrogram.max(axis=(1, 2), keepdims=True)
+        spectrogram = np.clip(spectrogram, a_min=max_values - db_range, a_max=None)
```
Comment (ylacombe) [inline src/transformers/audio_utils.py:886]: Looking clean :fire:  !

File: src/transformers/audio_utils.py [hunk 7]
```diff
@@ -680,6 +937,51 @@ def amplitude_to_db(
...
+
+    spectrogram = np.clip(spectrogram, a_min=min_value, a_max=None)
+    spectrogram = 20.0 * (np.log10(spectrogram) - np.log10(reference))
+
+    if db_range is not None:
+        if db_range <= 0.0:
+            raise ValueError("db_range must be greater than zero")
+        # Apply db_range clipping per batch item
+        max_values = spectrogram.max(axis=(1, 2), keepdims=True)
+        spectrogram = np.clip(spectrogram, a_min=max_values - db_range, a_max=None)
```
Comment (ylacombe) [inline src/transformers/audio_utils.py:980]: Great as well!

File: tests/utils/test_audio_utils.py [hunk 8]
```diff
@@ -20,19 +20,21 @@
...
+    spectrogram_batch,
     window_function,
 )
-from transformers.testing_utils import is_librosa_available, require_librosa
+from transformers.testing_utils import is_librosa_available
 
 
 if is_librosa_available():
-    from librosa.filters import chroma
+    pass
```
Comment (ylacombe) [inline tests/utils/test_audio_utils.py:37]: Why are we checking if librosa is available if we don't import anything at the end of the day ?

File: tests/utils/test_audio_utils.py [hunk 9]
```diff
@@ -762,56 +1606,80 @@ def test_amplitude_to_db(self):
         with pytest.raises(ValueError):
             amplitude_to_db(spectrogram, db_range=-80)
 
-    @require_librosa
-    def test_chroma_equivalence(self):
```
Comment (ylacombe) [inline tests/utils/test_audio_utils.py:766]: I think you removed this test, probably a mistake right ? In that case, let's put it back and reimport librosa !

File: tests/utils/test_audio_utils.py (hunk 8 above)
Comment (ylacombe) [reply tests/utils/test_audio_utils.py:37]: I think you removed a test by mistake !

File: tests/utils/test_audio_utils.py
Comment (ylacombe) [inline tests/utils/test_audio_utils.py:1]: Thanks so much for adding these thorough tests, looks like the method works as expected now!

**Review by ylacombe — APPROVED** [review]
Hey @ravenouse, I hope you're doing well!
First of all, I'm really sorry for the late review and thank you so much for working on this!

This looks great to me as it is! Let's ask for a core maintainer review!

In the meantime, I believe we should:
1. re-add the test that has been removed by mistake
2. conduct a speed benchmark as compared to the o.g method
 
Again, thanks for the great work here, this is a beautiful PR!

**Comment by ylacombe** [comment]
Benchmarks are looking great, thanks for the effort!

cc @amyeroberts, could you review this PR ?

File: src/transformers/audio_utils.py [hunk 10]
```diff
@@ -581,6 +581,214 @@ def spectrogram(
...
+        reference (`float`, optional, default=1.0):
+            Reference value for dB conversion in log_mel.
+        min_value (`float`, optional, default=1e-10):
+            Minimum floor value for log scale conversions.
+        db_range (`float`, optional):
+            Dynamic range for dB scale spectrograms.
+        remove_dc_offset (`bool`, optional):
+            Whether to remove the DC offset from each frame.
+        dtype (`np.dtype`, optional, default=np.float32):
+            Data type of the output spectrogram.
```
Comment (amyeroberts) [inline src/transformers/audio_utils.py:673]: nits 

```suggestion
        waveform_list (`List[np.ndarray]` with arrays of shape `(length,)`):
            The list of input waveforms, each a single-channel (mono) signal.
        window (`np.ndarray` of shape `(frame_length,)`):
            The windowing function to apply, including zero-padding if necessary.
        frame_length (`int`):
            The length of each frame for analysis.
        hop_length (`int`):
            The step size between successive frames.
        fft_length (`int`, *optional*):
            The size of the FFT buffer, defining frequency bin resolution.
        power (`float`, *optional*, defaults to 1.0):
            Determines the type of spectrogram: 1.0 for amplitude, 2.0 for power, None for complex.
        center (`bool`, *optional*, defaults to `True`):
            Whether to center-pad the waveform frames.
        pad_mode (`str`, *optional*, defaults to `"reflect"`):
            The padding strategy when `center` is `True`.
        onesided (`bool`, *optional*, defaults to `True`):
            If True, returns a one-sided spectrogram for real input signals.
        preemphasis (`float`, *optional*):
            Applies a pre-emphasis filter to each frame.
        mel_filters (`np.ndarray`, *optional*):
            Mel filter bank for converting to mel spectrogram.
        mel_floor (`float`, *optional*, defaults to 1e-10):
            Floor value for mel spectrogram to avoid log(0).
        log_mel (`str`, *optional*):
            Specifies log scaling strategy; options are None, "log", "log10", "dB".
        reference (`float`, *optional*, defaults to 1.0):
            Reference value for dB conversion in log_mel.
        min_value (`float`, °optional*, defaults to 1e-10):
            Minimum floor value for log scale conversions.
        db_range (`float`, *optional*):
            Dynamic range for dB scale spectrograms.
        remove_dc_offset (`bool`, *optional*):
            Whether to remove the DC offset from each frame.
        dtype (`np.dtype`, *optional*, defaults to `np.float32`):
            Data type of the output spectrogram.
```

File: src/transformers/audio_utils.py [hunk 11]
```diff
@@ -581,6 +581,214 @@ def spectrogram(
...
+            np.pad(
+                waveform,
+                padding,
+                mode=pad_mode,
+            )
+            for waveform in waveform_list
+        ]
+    original_waveform_lengths = [
+        len(waveform) for waveform in waveform_list
+    ]  # these lengths will be used to remove padding later
```
Comment (amyeroberts) [inline src/transformers/audio_utils.py:715]: nit - library convention is comments on the line above to avoid line splitting like this 

```suggestion
    # these lengths will be used to remove padding later
    original_waveform_lengths = [len(waveform) for waveform in waveform_list]
```

File: src/transformers/audio_utils.py [hunk 12]
```diff
@@ -581,6 +581,214 @@ def spectrogram(
...
+
+    # Promote to float64, since np.fft uses float64 internally
+    padded_waveform_batch = padded_waveform_batch.astype(np.float64)
+    window = window.astype(np.float64)
+
+    # Split waveform into frames of frame_length size
+    num_frames = int(1 + np.floor((padded_waveform_batch.shape[1] - frame_length) / hop_length))
+    true_num_frames = [
+        int(1 + np.floor((length - frame_length) / hop_length)) for length in original_waveform_lengths
+    ]  # these lengths will be used to remove padding later
```
Comment (amyeroberts) [inline src/transformers/audio_utils.py:735]:
```suggestion
    # these lengths will be used to remove padding later
    true_num_frames = [int(1 + np.floor((length - frame_length) / hop_length)) for length in original_waveform_lengths]
```

**Review by amyeroberts — APPROVED** [review]
Nice 🔥 !

Very clean and easy to follow code - thanks for adding in such extensive tests too ❤️

**Comment by ylacombe** [comment]
Thanks again for this @ravenouse, let's discuss on how to integrate this on some feature extractors as a follow-up?  Would you like to give it a try?

**Comment by ylacombe** [comment]
Hey @ravenouse, thanks for following that up!
The plan sounds good! 
There's only a [few models](https://github.com/search?q=repo%3Ahuggingface%2Ftransformers%20_extract_fbank_features&type=code) that are using `_extract_fbank_features`, let's do them all at once?
If you feel more comfortable starting with Seamless M4T, don't hesitate as well.  Don't hesitate to ask for a review early on as well!
